src/main/java/com/hopper/cloud/airlines/HopperClient.java

```diff
@@ -165,6 +169,57 @@ public CfarContract getContract(String sessionId, String contractId) throws ApiE
         return cfarApi.getCfarContractsId(contractId, sessionId);
     }
 
+    /**
+     * Called to update forms of payment that are linked with a contract.
+     *
+     * @param sessionId                         The current session IO
+     * @param contractReference                 The contract reference
+     * @param updateCfarFormOfPaymentRequest    The request with the update information
+     * @return The updated contract
+     * @throws ApiException
+     */
+    public CfarContract updateCfarContractFormsOfPayment(String sessionId, String contractReference, UpdateCfarFormOfPaymentRequest updateCfarFormOfPaymentRequest) throws ApiException {
+        try {
+            if (ListUtil.isEmpty(updateCfarFormOfPaymentRequest.getFormsOfPayment())) {
+                throw new ApiException("Missing forms of payment"); // must we return the contract instead?
+            } else {
+                if (hopperPaymentClient == null) {
+                    throw new ApiException("Missing credentials for payment");
+                }
+                // Tokenize each credit card
+                List<FormOfPayment> updatedFormsOfPaymentRequest = new ArrayList<>();
+                for (FormOfPayment formOfPaymentRequest : updateCfarFormOfPaymentRequest.getFormsOfPayment()) {
+                    if (formOfPaymentRequest instanceof FormOfPayment.PaymentCard) {
+                        FormOfPayment.PaymentCard creditCardRequest = (FormOfPayment.PaymentCard)formOfPaymentRequest;
+                        if (creditCardRequest.getToken() == null) {
+                            // Retrieve the required token from Spreedly
+                            CreditCardDetail creditCardDetail = creditCardRequest.getCreditCardDetail();
+                            if (creditCardDetail == null) {
+                                throw new ApiException("Missing required details for a credit card");
+                            }
+                            // Adjust the credit card number
+                            creditCardDetail.setNumber(prepareCreditCardNumberForSpreedly(creditCardDetail.getNumber()));
+                            String token = hopperPaymentClient.tokenizePaymentCreditCardWithEncryption(creditCardDetail);
+
+                            // Build a new instance only with the token
+                            FormOfPayment.PaymentCard updatedCreditCardRequest = new FormOfPayment.PaymentCard(creditCardRequest.getAmount(), creditCardRequest.getCurrency(), token);
+                            updatedFormsOfPaymentRequest.add(updatedCreditCardRequest);
+                        } else {
+                            creditCardRequest.setCreditCardDetail(null);  // Credit card details must not be passed to the HTSFA API
```

Comment:
This is not very clean but this will make the job (ie - Do not pass credit card details  to the HTSFA API)
We could avoid to do this if we have made the choice to create two separated `UpdateCfarFormOfPaymentRequest` objects : the first one defined as input in this method and the second one used as input in the function provided by the cfarApi.
But there will be too much similarities between both models so finally,  it wasn’t worth it

Reply:
I disagree, but I don't know java conventions so please push back.

It seems valuable to use a separate model to guarantee that we're not sending the wrong field to the wrong place. It's a bit more boilerplate but the increased protection against the data leaking seems worth it. We really don't want accidental CC info showing up in our logs, for example.

Additionally, it doesn't seem like a good idea to mutate the caller's copy of an object. If I were calling this method, I wouldn't expect my creditCardDetail to suddenly be "consumed" and disappear.

Reply:
Ok ! I'll finally create a separate (and specific) model, so as to distinguish the request object between the Java SDK and the HTFSA API